src/charm.py

```diff
@@ -305,180 +142,70 @@ def _on_relation_departed(self, event: RelationDepartedEvent, charm_state: Charm
+        """Reconcile Synapse configuration with charm state.
 
-    def set_signing_key(self, signing_key: str) -> None:
-        """Create secret with signing key content.
+        This is the main entry for changes that require a restart.
 
         Args:
-            signing_key: signing key as string.
+            charm_state: Instance of CharmState
         """
-        peer_relation = self.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
-        if not peer_relation:
-            logger.error(
-                "Failed to set signing key: no peer relation %s found",
-                synapse.SYNAPSE_PEER_RELATION_NAME,
-            )
+        container = self.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
             return
-
-        if signing_key == self.get_signing_key():
-            logger.info("Received signing key but there is no change, skipping")
+        if helpers.is_redis_required(self, charm_state):
+            self.unit.status = ops.BlockedStatus("Redis integration is required.")
             return
+        self.model.unit.status = ops.MaintenanceStatus("Configuring Synapse")
+        helpers.configure_and_start_services(self, charm_state, container)
+        if helpers.is_mjolnir_enabled(self, charm_state):
+            self._mjolnir.enable(charm_state)
         if self.unit.is_leader():
-            logger.debug("Adding signing key to secret: %s", signing_key)
-            secret = self.app.add_secret({"secret-signing-key": signing_key})
-            peer_relation[0].data[self.app].update(
-                {"secret-signing-id": typing.cast(str, secret.id)}
-            )
-
-    def get_signing_key(self) -> typing.Optional[str]:
-        """Get signing key from secret.
+            self._matrix_auth.update_matrix_auth_integration(charm_state)
+        self._set_unit_status()
 
-        Returns:
-            Signing key as string or None if not found.
-        """
-        peer_relation = self.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
-        if not peer_relation:
-            logger.error(
-                "Failed to get signing key: no peer relation %s found",
-                synapse.SYNAPSE_PEER_RELATION_NAME,
-            )
-            return None
-
-        secret_id = peer_relation[0].data[self.app].get("secret-signing-id")
-        if secret_id:
-            try:
-                secret = self.model.get_secret(id=secret_id)
-                logging.debug(secret.get_content().get("secret-signing-key"))
-                return secret.get_content().get("secret-signing-key")
-            except (ops.model.SecretNotFoundError, ValueError, TypeError) as exc:
-                logger.exception("Failed to get secret id %s: %s", secret_id, str(exc))
-                del peer_relation[0].data[self.app]["secret-signing-id"]
-        return None
+    def _set_unit_status(self) -> None:
```

Comment:
I think that this method is not necessary anymore, `reconcile` is pretty short and already setting some status. This method is only called from reconcile, and contains some overlapping code (`can_connect` for example)